arangod/Aql/ConstrainedSortExecutor.cpp

```diff
@@ -0,0 +1,257 @@
+  using arangodb::aql::AqlItemBlock;
+  using arangodb::aql::AqlValue;
+  using arangodb::aql::RegisterId;
+
+  size_t dRow = _rowsPushed;
+
+  if (_rowsPushed >= _infos._limit) {
+    // pop an entry first
+    std::pop_heap(_rows.begin(), _rows.end(), *_cmpHeap.get());
+    dRow = _rows.back();
+     eraseRow(*_heapBuffer, dRow);
+    _rows.pop_back();
+  }
+  TRI_ASSERT(dRow < _infos._limit);
+
+  TRI_IF_FAILURE("SortBlock::doSortingInner") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  _heapOutPutRow.*get(RowTag()) = dRow;
+  _heapOutPutRow.copyRow(input, true);
+  _heapOutPutRow.advanceRow();
+
+  _rows.emplace_back(dRow); // add to heap vector
+  LOG_DEVEL << "increase rows pushed";
+  ++_rowsPushed;
+
+  // now restore heap condition
+  std::push_heap(_rows.begin(), _rows.end(), *_cmpHeap.get());
+  LOG_DEVEL << "exit push row" << _rowsPushed;
+
+  return TRI_ERROR_NO_ERROR;
+}
+
+bool ConstrainedSortExecutor::compareInput(uint32_t const& rowPos, InputAqlItemRow& row) const {
+  for (auto const& reg : _infos.sortRegisters()) {
+    auto const& lhs = _heapBuffer->block().getValueReference(rowPos, reg.reg);
+    auto const& rhs = row.getValue(reg.reg);
+
+    int const cmp = arangodb::aql::AqlValue::Compare(_infos.trx(), lhs, rhs, true);
+
+    if (cmp < 0) {
+      return reg.asc;
+    } else if (cmp > 0) {
+      return !reg.asc;
+    }
+  }
+  return false;
+}
+
+ConstrainedSortExecutor::ConstrainedSortExecutor(Fetcher& fetcher, SortExecutorInfos& infos)
+    : _infos(infos),
+      _fetcher(fetcher),
+      _returnNext(0),
+      _outputPrepared(false),
+      _rowsPushed(0),
+      _heapBuffer(std::make_shared<AqlItemBlockShell>(_infos._manager, std::unique_ptr<AqlItemBlock>(_infos._manager.requestBlock(_infos._limit, _infos.numberOfOutputRegisters())))),
+      _cmpHeap(std::make_unique<ConstrainedLessThan>(_infos.trx(), _infos.sortRegisters())),
+      _heapOutPutRow{_heapBuffer, infos.getOutputRegisters(), infos.registersToKeep()},
+      _done(false)
+  {
+  TRI_ASSERT(_infos._limit > 0);
+  _rows.reserve(infos._limit);
+  _cmpHeap->setBuffers(&_heapBuffer->block(), &_heapBuffer->block());
+  LOG_DEVEL << "####################### CREATED EXECUTRO ################################";
+};
+
+ConstrainedSortExecutor::~ConstrainedSortExecutor() = default;
+
+std::pair<ExecutionState, NoStats> ConstrainedSortExecutor::produceRow(OutputAqlItemRow& output) {
+  if(_done){
+    return {ExecutionState::DONE, NoStats{}};
+  }
+
+  ExecutionState state = ExecutionState::HASMORE;
+
+  if (!_outputPrepared) {
+    // build-up heap by pulling form all rows in the singlerow fetcher
+    state = doSorting();
+  }
+
+  if (state != ExecutionState::HASMORE) {
+    TRI_ASSERT(state != ExecutionState::WAITING || state == ExecutionState::DONE);
+    // assert heap empty
+    return {state, NoStats{}};
+  }
+
+  TRI_ASSERT( state == ExecutionState::HASMORE && _outputPrepared);
+  std::sort(_rows.begin(), _rows.end(), *_cmpHeap);
```

Comment:
Move this into `doSorting` so it's only executed once.

Reply:
Yes! I thought you would find it:) That one and that I will always allocate the limit elements. I need to re-introduce the ensureHeapBuffer.

Reply:
Well, the `ensureHeapBuffer` was used initially because I couldn't allocate it in the constructor yet. In the old pattern, we didn't have the number of registers until we got the first source block, as far as I could tell. (I may have missed it.) But that was the only reason I didn't allocate the heap buffer in the constructor.

Reply:
I have seen some times that the block returns immediately DONE. Memory allocation could be avoided in such cases. But It is not the regular case :)